### DISCUSSION TITLE
[A/L] (2021/22) Funzioni - Esercizio 8

### DISCUSSION THREAD

**Discussion by matypist** [sapienzastudentsnetwork/metodi-matematici-per-l-informatica#248] [Esercizi A-L]
![image](https://user-images.githubusercontent.com/69389461/152315257-1a5fa43d-f1a2-4f97-b282-9b221eb759fc.png)

**Comment by GattoBrizzolato** [comment]
![es8](https://user-images.githubusercontent.com/98970164/172479803-c4576f06-d4bd-447c-a97a-e9bb6b88f1da.jpeg)

**Comment by LuginiAndrea** [comment]
Ad ogni elementi dell'insieme di partenza posso associare uno dei 4 elementi dell'insieme di arrivo. 
Quindi cardinalità di $F = 4  \times 4 \times 4 \times 4 \times 4 = 4^5$
Ora, per buona traduzione possiamo costruire una stringa che associa ad ogni elemento dell'insieme di partenza un elemento del suo insieme di arrivo. Possiamo facilmente vedere come questa definizione della stringa equivalga alla funzione.
Possiamo quindi costruire 4 classi di stringhe: $11234,12234,12334,12344$.
In quanto l'ordine conta, possiamo ora considerare ogni anagramma di queste stringhe = $\frac{5!}{2!} \times 4 = 240$ possibili funzioni suriettive.
Infine, non può esistere una funzione iniettiva, in quanto la cardinalità dell'insieme di partenza è maggiore della cardinalità dell'insieme di arrivo, e di conseguenza non può esistere nessuna funzione biettiva

**Comment by Elia-Belli** [comment]
Siano A ={a,b,c,d,e} e B ={1,2,3,4}

**Funzioni Iniettive**
E' evidente che non sia possibile costruire funzioni iniettive da A in B in quanto non sono presenti sufficienti elementi in B per avere un'immagine unica per ogni elemento di A.

Segue che non esistono **funzioni biettive**.

**Comment by LuginiAndrea** [comment]
> Siano A ={a,b,c,d,e} e B ={1,2,3,4}
> 
> **Funzioni Iniettive** E' evidente che non sia possibile costruire funzioni iniettive da A in B in quanto non sono presenti sufficienti elementi in B per avere un'immagine unica per ogni elemento di A.
> 
> Segue che non esistono **funzioni biettive**.
> 
> **Funzioni Suriettive** Possiamo contare le funzioni suriettive come se stessimo risolvendo un problema di Scrittura Additiva con una condizione di almeno 1. Quindi distribuiamo 1 elemento di A ai 4 elementi di B, siccome A è un insieme i suoi elementi sono differenti e dobbiamo contare in quanti modi possiamo distribuire questi elementi (disposizioni semplici: 4!). Infine dobbiamo distribuire l'ultimo elemento rimasto, il chè può avvenire solamente in 4 modi. Quindi le funzioni suriettive sono: 4 * 4! = 96

Il ragionamento che esegui nel secondo punto è sbagliato:
la tua scrittura additiva è chiaramente nella forma
$$x_1 + x_2 + x_3 + x_4 = 5$$
Il problema nel tuo ragionamento è che ignori _quali_ elementi di A distribuiamo, dicendo solo che abbiamo 4! modi per distribuirli.
Dobbiamo inanzitutto scegliere quali elementi distribuire: 
$$\binom{5}{4}$$
Ora quanti modi abbiamo di distribuirli: $4!$
Ne consegue che abbiamo $5 \cdot 4! = 5! = 120$ modi di distribuire 4 elementi di A
Rimane ora la questione su come distribuire l'ultimo: 4 modi.
Bisogna però tenere in considerazione che stiamo contando il doppio delle volte: difatti l'ordine di distribuzione non conta, e visto che c'è almeno un $x$ a cui vengono distribuiti due elementi uguali, vuol dire che contiamo il doppio delle volte.
Ne consegue che il numero di distribuzioni sono:
$$\binom{5}{4} \cdot 4! \cdot 4 \cdot \frac{1}{2}$$
Ad ulteriore conferma di questo risultato è la formula che usa i numeri di Sterling del secondo tipo:
https://math.stackexchange.com/questions/2866173/number-of-surjective-functions-from-a-set-with-m-elements-onto-a-set-with-n

**Comment by IonutCicio** [comment]
Di seguito una soluzione generica a questa domanda (con i due insiemi di cardinalità $n$ e $k$) https://github.com/sapienzastudentsnetwork/calcolo-delle-probabilita/discussions/153

Ponendo i valori $n = 5$ e $k = 4$ ne risulta che il numero di funzioni suriettive è uguale a

$$
\sum_{i = 0}^{4} (-1)^i\binom{4}{i}(4 - i)^5 =
$$

$$
= { 4 \choose 0 } \cdot 4^5 - { 4 \choose 1 } \cdot 3^5 + { 4 \choose 2 } \cdot 2^5 - { 4 \choose 3 } \cdot 1^5 + { 4 \choose 4 } \cdot 0^5 =
$$

 $$
 = 4^5 - 4 \cdot 3^5 + 6 \cdot 2^5 - 4 = 240
 $$
 
 Lo stesso risultato del collega @LuginiAndrea  (la sua rimane comunque una soluzione più elegante in questo caso)
